service/history/workflow/mutable_state_impl.go

```diff
@@ -1350,12 +1351,47 @@ func (ms *MutableStateImpl) GetCompletionEvent(
 		currentBranchToken,
 	)
 	if err != nil {
+		if (common.IsNotFoundError(err) ||
+			common.IsInternalError(err)) &&
+			ms.executionState.Status == enumspb.WORKFLOW_EXECUTION_STATUS_TERMINATED {
+			// Certain terminated workflows have an incorrect completionEventBatchId recorded which
+			// prevents the completion event from being loaded outside of cache.
+			//
+			// If we get back an internal error, attempt to search history (most recent
+			// events first) to find the completion event. Event will be earlier than
+			// the recorded CompletionEventBatchID, and should be a part of the same batch.
+			//
+			// See also: https://github.com/temporalio/temporal/pull/6180
+			//
+			// TODO: Remove 90 days after deployment (remove after 10/16/2024)
+			_, txID := ms.GetLastFirstEventIDTxnID()
+			resp, err := ms.shard.GetExecutionManager().ReadHistoryBranchReverse(ctx, &persistence.ReadHistoryBranchReverseRequest{
+				ShardID:                ms.shard.GetShardID(),
+				BranchToken:            currentBranchToken,
+				MaxEventID:             nextEventID, // looking for an event in the most recent batch
+				PageSize:               1,
+				LastFirstTransactionID: txID,
+				NextPageToken:          []byte{},
+			})
+			if err != nil {
+				return nil, err
```

Comment:
let's catch the not found error from here as well. probably make it a defer function.

Reply:
done